### PR TITLE
OSD-16469 Block ingress controller provisioning on master and infra

### PR DESCRIPTION
**Pull request by Tafhim**
This PR adds a webhook that inspects ingress controller creation request from customers. It blocks these requests when they have tolerations for infra and master nodes.

Issue: [OSD-16469](https://issues.redhat.com//browse/OSD-16469)

### CODE REVIEW

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (bmeng) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:115]: I am not sure if this required, as the operator itself will reconcile the `ingresscontroller` within the namespace only, so created outside should be harmless but doing nothing.

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (bmeng) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:122]: This as well, the default ingresscontroller should be managed by tools. I am not quite sure if this is needed.

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (bmeng) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:26]: This should be namespace scope?

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (bmeng) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:29]: We should limited to CREATE/UPDATE/PATCH here.

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (bmeng) [reply pkg/webhooks/ingresscontroller/ingresscontroller.go:122]: Also we will need to allow the default ingresscontroller to be created by special user/group, so that it can be created after sync from ocm/hive.

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (hectorakemp) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:115]: Maybe also a loop here over `nodeSelector` labels?
I.e. reject yaml like this:
```
      nodeSelector:
        node-role.kubernetes.io/infra: "" #or master
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/infra
          operator: Exists
```

**Review by hectorakemp — CHANGES_REQUESTED** [review]
99% there - amazing work @Tafhim in such a short space of time. One small addition to the authorize loop, and then I think we're there.

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (hectorakemp) [reply pkg/webhooks/ingresscontroller/ingresscontroller.go:115]: Resolving from discussion here: https://redhat-internal.slack.com/archives/C0326L38PEH/p1686795999616689

Comment (bmeng) [inline pkg/webhooks/ingresscontroller/ingresscontroller.go:115]: will this block the operators to update the ingresscontroller?

**Comment by bmeng** [comment]
/hold 
Hold for now till tomorrow, just to unblock the upcoming promotion, ready for review

File: pkg/webhooks/ingresscontroller/ingresscontroller.go
Comment (hectorakemp) [reply pkg/webhooks/ingresscontroller/ingresscontroller.go:115]: I don't think it will @bmeng, because operators don't ever deploy workloads to non-infra or master nodes, do they?

Comment (hectorakemp) [reply pkg/webhooks/ingresscontroller/ingresscontroller.go:118]: I don't think we do @Tafhim - if all three of those operators are able to create ICs, then we're good

**Review by hectorakemp — APPROVED** [review]

**Pull request merged**
